### DISCUSSION TITLE
About the vertical parallelization of Processors

### DISCUSSION THREAD

**Discussion by JasonLi-cn** [databendlabs/databend#8117] [Q&A] [answered]
https://github.com/datafuselabs/databend/blob/4bf7812f315d5ec6a1720e55ef7fb6b8d3d32823/src/query/pipeline/core/src/pipeline.rs#L116-L123

Whether (max_threads <= max_pipe_size) affects the vertical parallelization of Processors?

BTW:
https://github.com/datafuselabs/databend/blob/4bf7812f315d5ec6a1720e55ef7fb6b8d3d32823/src/query/pipeline/core/src/processors/port.rs#L169-L184

If unset NEED_DATA here, the OutputPort::can_push() function will return false when the upstream ProcessorPtr calls event(), then the upstream ProcessorPtr will enter State::Idle in schedule_queue() function and will not be executed by worker. Thus, vertical parallel execution fails.

REF:
<img width="1173" alt="image" src="https://user-images.githubusercontent.com/23028310/195060670-d47f3456-e9ef-4b51-9261-c2293904b538.png">

**Comment by zhang2014** [comment, marked as answer]
This is determined by task stealing. vertical execution will occur only when has idle thread.
